### PR TITLE
Implement assembler modifiers

### PR DESCRIPTION
**Pull request by agnat**
`hi5(...)` et al ;)

### CODE REVIEW

**Comment by agnat** [comment]
This is work in progress. Do not merge until further notice. ;)

**Comment by agnat** [comment]
I doubt any reasonable C++ compiler actually implements arithmetic shifting. In combination with promotion to _signed_ integers (bitwise or, &c.) it's just to dangerous ... and _never_ what you want. But I'm not an expert in these things...

**Comment by Sh4rK** [comment]
If you do
```c++
int32_t x = 0xff000000;
x = x >> 24;
```
then it actually performs an arithmetic shift (the result is 0xffffffff), so I really don't know what's going on.

**Comment by agnat** [comment]
Hehe, that's why it's called "implementation defined behavior" :P

I'm also pretty sure the literal `0xff000000` is actually an unsigned... anyway...

**Comment by Sh4rK** [comment]
Then I guess we'll never now :D

**Comment by agnat** [comment]
Ready for review.

@dylanmckay, could you look into adding the other fixups around [here](https://github.com/agnat/llvm/blob/feature/assembler_operators/lib/Target/AVR/MCTargetDesc/AVRMCCodeEmitter.cpp#L255)?

File: lib/Target/AVR/AsmParser/AVRAsmParser.cpp [hunk 1]
```diff
@@ -281,7 +235,7 @@ MatchAndEmitInstruction(SMLoc IDLoc, unsigned &Opcode,
 
 int AVRAsmParser::parseRegister() {
   const AsmToken &Tok = Parser.getTok();
-  int RegNum = -1;
+  int RegNum = 0;
```
Comment (dylanmckay) [inline lib/Target/AVR/AsmParser/AVRAsmParser.cpp]: I'm not sure if this will work in all cases. Isn't `r0

Comment (agnat) [inline lib/Target/AVR/AsmParser/AVRAsmParser.cpp]: Pretty sure it's ok. `NoRegister` is zero and I looked at the generated function `MatchRegisterName(...)`. But good looking out there ;)

**Comment by dylanmckay** [comment]
The PR looks really nice! (it also makes me realise how much cleaner my `switch` statements would look if I condensed them up :P).

If you want to roll the commits into one or two commits, I'll merge :)

**Comment by agnat** [comment]
Squashed.

**Comment by dylanmckay** [comment]
I'll work on the fixup code tomorrow.

**Pull request merged**